### PR TITLE
fix(NODE-3712,NODE-4546): electionId should be ordered before setVersion

### PR DESCRIPTION
**Pull request by nbbeeken**
### Description

NODE-3712, NODE-4546

#### What is changing?

- Sync spec tests, which suffciently unit test the changes
- Make the logical changes to the `updateRsFromPrimary` procedure
- Update the unit test runner, debugability / typescript / modern code
  - Tests passed before and after the changes here
  
Spec runner changes and partial spec sync done in #3182 
- description update in: secondary_ignore_ok_0.json (https://github.com/mongodb/specifications/commit/5bd06a81d97850a1365d73f08bbb7dea562e72b9)

Update Oct 2022:
* Includes the condition to only use the new behavior on wireVersions 17 or greater

#### What is the motivation for this change?

Correctness, see spec changes: https://github.com/mongodb/specifications/pull/1122
See follow up spec changes unblocking this: https://github.com/mongodb/specifications/pull/1309

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/sdam/topology_description.ts [hunk 1]
```diff
@@ -362,25 +362,29 @@ function topologyTypeForServerType(serverType: ServerType): TopologyType {
   }
 }
 
-// TODO: improve these docs when ObjectId is properly typed
-function compareObjectId(oid1: Document, oid2: Document): number {
+function compareObjectId(oid1?: ObjectId, oid2?: ObjectId): 0 | 1 | -1 {
```
Comment (durran) [inline src/sdam/topology_description.ts:365]: Just my feeling here that `compareObjectId` and `compareNumber` might better be placed in the utils module as they are pretty generic.

**Review by durran — CHANGES_REQUESTED** [review]

**Review by dariakp — CHANGES_REQUESTED** [review]

**Review by dariakp — CHANGES_REQUESTED** [review]

File: test/unit/assorted/server_discovery_and_monitoring.spec.test.ts [hunk 2]
```diff
@@ -0,0 +1,382 @@
+import { Document, EJSON } from 'bson';
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/unit/assorted/server_discovery_and_monitoring.spec.test.ts:1]: #3182 Done! you have to go in and view the second commit to see the diff as opposed to whole file delete and rewrite. I'll push the removal from here in a bit

File: src/sdam/topology_description.ts [hunk 3]
```diff
@@ -362,25 +362,29 @@ function topologyTypeForServerType(serverType: ServerType): TopologyType {
...
   if (oid2 == null) {
     return 1;
   }
 
-  if (oid1.id instanceof Buffer && oid2.id instanceof Buffer) {
-    const oid1Buffer = oid1.id;
-    const oid2Buffer = oid2.id;
-    return oid1Buffer.compare(oid2Buffer);
+  const res = oid1.id.compare(oid2.id);
+  return res === 0 ? 0 : res > 0 ? 1 : -1;
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/sdam/topology_description.ts]: Yup fixed, before my change we were using .compare on the hex strings and javascript's string .compare isn't guaranteed to return `1|-1|0`, but node buffer compare does!

File: src/sdam/topology_description.ts [hunk 4]
```diff
@@ -362,25 +362,29 @@ function topologyTypeForServerType(serverType: ServerType): TopologyType {
...
+}
+
+function compareNumber(num1?: number, num2?: number): 0 | 1 | -1 {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/sdam/topology_description.ts]: removed.

File: src/sdam/topology_description.ts [hunk 5]
```diff
@@ -396,31 +400,30 @@ function updateRsFromPrimary(
...
 
-        return [checkHasPrimary(serverDescriptions), setName, maxSetVersion, maxElectionId];
-      }
-    }
-
-    maxElectionId = serverDescription.electionId;
-  }
+  const setVersionComparison = compareNumber(maxSetVersion, serverDescription.setVersion);
+  const maxSetVersionIsLess = setVersionComparison === -1;
+  const maxSetVersionIsEqual = setVersionComparison === 0;
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/sdam/topology_description.ts]: Take a look at the updated logic, my verboseness was encouraged by the debugging that was done during authoring the spec, I think my simplification captured what you're looking for, and I agree it's just as readable/intuitive.

File: test/unit/assorted/server_discovery_and_monitoring.spec.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dariakp) [reply test/unit/assorted/server_discovery_and_monitoring.spec.test.ts]: Leaving this comment open so we can remember to rebase once the other one is in

File: src/sdam/topology_description.ts (hunk 1 above)
Comment (nbbeeken) [reply src/sdam/topology_description.ts]: `compareObjectId` was moved to utils in #3348, rebased.

File: src/utils.ts [hunk 6]
```diff
@@ -1300,7 +1300,7 @@ export function getMongoDBClientEncryption(): {
  * - `-1 = oid1 is less than oid2`
  * - `+0 = oid1 is equal oid2`
  */
-export function compareObjectId(oid1?: ObjectId, oid2?: ObjectId): 0 | 1 | -1 {
+export function compareObjectId(oid1?: ObjectId | null, oid2?: ObjectId | null): 0 | 1 | -1 {
```
Comment (baileympearson) [inline src/utils.ts:1303]:
```suggestion
export function compareObjectId(oid1: ObjectId | null, oid2: ObjectId | null): 0 | 1 | -1 {
```
will oid1 / oid2 ever be undefined?

**Review by baileympearson — APPROVED** [review]
optional comment, otherwise LGTM

File: src/utils.ts (hunk 6 above)
Comment (nbbeeken) [reply src/utils.ts:1303]: It won't be in our usage but the function would support those inputs cus of the way we check nullish. I've added tests to keep it this way.

**Review by baileympearson — APPROVED** [review]
CI succeeded on restart, just flakey tests.

**Pull request merged**